File: LC/java/Reverse Nodes in k-Group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if(head == null || k <= 1){
            return head;
        }
        ListNode start,end,h;
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        start = dummy;
        h = head;
        int cnt = 0;
        
        while(h!=null){
            cnt += 1;
            end = h.next;
            if (cnt == k){
                start = reversePart(start,end);    
                cnt = 0;
            }
            h = end;
        }
        return dummy.next;
        
    }
    
    public ListNode reversePart(ListNode start, ListNode end){
        ListNode a,b,c;
        a = start.next;
        b = a.next;
        
        if (start == null || start.next == null){
            return start;
        }
        
        while(b != end){
            c = b.next;
            b.next = start.next;
            start.next = b;
            b = c;
        }
        a.next = end;
        return a;
    }
}

File: LC/python/Reverse Nodes in k-Group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @param k, an integer
    # @return a ListNode
    def reverseKGroup(self, head, k):
        if head == None or k<=1:
            return head
        dummy = ListNode(0)
        dummy.next = head
        start = dummy
        
        while start.next:
            end = start
            for i in range(k-1):
                end = end.next
                if end.next == None:
                    return dummy.next
            r = self.reversePart(start.next,end.next)
            start.next = r[0]
            start = r[1]
        return dummy.next
        
        
    def reversePart(self,start,end):
        d = ListNode(0)
        d.next = start
        
        while d.next != end:
            tmp = start.next
            start.next = tmp.next
            tmp.next = d.next 
            d.next = tmp
        return [end,start]
        
